File: packages/@cos/extension-points/src/stub-base.ts
// Extension Point Stub Base — TypeScript
// Source: context/00_master_construction_os.md §EXTENSION POINT PROTOCOL
// Every extension_point() stub MUST extend this class and call logStubCall().

import { createLogger } from '@cos/logger';

const epLogger = createLogger('extension-points');

export abstract class StubBase {
  abstract readonly EP_ID: string;
  abstract readonly EP_VERSION: string;
  abstract readonly TRIGGER: string;
  abstract readonly PHASE: string;

  /**
   * Must be called at the start of every stub method.
   * Emits a WARN log so stubs are always visible in observability.
   * QM-8: trace_id is injected when available via AsyncLocalStorage context.
   */
  protected logStubCall(methodName: string, context?: Record<string, unknown>): void {
    epLogger.warn(
      {
        ep_id: this.EP_ID,
        ep_version: this.EP_VERSION,
        phase: this.PHASE,
        trigger: this.TRIGGER,
        method: methodName,
        context,
      },
      `[STUB] ${this.EP_ID}#${methodName} called — implement when: ${this.TRIGGER}`,
    );
  }
}

File: ai/shared/stub_base.py
"""
Extension Point Stub Base — Python
Source: context/00_master_construction_os.md §EXTENSION POINT PROTOCOL
Every extension_point() stub MUST extend this class and call log_stub_call().
"""

import logging
from abc import ABC, abstractmethod
from typing import Any

logger = logging.getLogger("cos.extension-points")


class StubBase(ABC):
    @property
    @abstractmethod
    def EP_ID(self) -> str: ...

    @property
    @abstractmethod
    def EP_VERSION(self) -> str: ...

    @property
    @abstractmethod
    def TRIGGER(self) -> str: ...

    @property
    @abstractmethod
    def PHASE(self) -> str: ...

    def log_stub_call(self, method_name: str, context: dict[str, Any] | None = None) -> None:
        """Must be called at the start of every stub method — emits WARN so stubs are visible."""
        logger.warning(
            "[STUB] %s#%s called — implement when: %s",
            self.EP_ID,
            method_name,
            self.TRIGGER,
            extra={
                "ep_id": self.EP_ID,
                "ep_version": self.EP_VERSION,
                "phase": self.PHASE,
                "method": method_name,
                "context": context or {},
            },
        )
